### PR TITLE
Issue194 slow undo

### PR DESCRIPTION
**Pull request by gthazmatt**
# PullRequest

Great that you are thinking about submitting a PullRequest. Please keep them small if not otherwise possible. Contact me *before* you are creating a big PR to avoid unnecessary work and rebasing of big PRs. Also try to add tests - I am not religious about that but prefer PRs backed by tests. This project has everything setup for Espresso UI and Unit-tests. Also the existing unit and UI-tests muss pass before submitting a PullRequest. Please base the PullRequests on dev.
# PullRequest

Great that you are thinking about submitting a PullRequest!

Please keep them small if not otherwise possible. Contact me *before* you are c\
reating a big PR to avoid unnecessary work and rebasing of big PRs. Also try to\
 add tests - I am not dogmatic about that but prefer PRs backed by tests. This \
 project has everything setup for Espresso UI and Unit-tests. Also the existing \
 unit and UI-tests muss pass before submitting a PullRequest.

Please base the PullRequests on the branch named dev

### CODE REVIEW

**Comment by ligi** [comment]
Thanks for the PR! Looks good - but will have to have a deeper look later - very late here. One UI test is failing on jenkins(We Can Undo Pass TheUndoFunctionality) - is it passing on your end?

**Comment by gthazmatt** [comment]
I didn't see any indication of it failing, but I'm not as familiar with gradle. It did occur to me while I was in the shower that I forgot to check pass moves, so there's a good chance that I missed something. I also don't have much experience with android, so 

I put this PR out a little early because I won't be able to look at it again for a few days, and I expected you would want some changes or additions.

**Comment by gthazmatt** [comment]
Damn it. I hate doing crap on my phone. I was editing my message and managed to hit the close button. The first paragraph above was supposed to end with "so, I haven't figured out how to do some things like run a test or redeploy without rebuilding everything."

**Comment by gthazmatt** [comment]
Oh, I also have a bad habit of wanting to fix everything all at once, especially things like inconsistent formatting and spacing which really screws with diffs. I tried to keep things isolated, but if you want me to back out anything, just let me know.

**Comment by ligi** [comment]
Thanks for the update - UI-tests now pass - but there are lint errors - they need to be fixed

**Comment by gthazmatt** [comment]
I don't think the tests are running when I build, and if I try to run a test class directly, it says "Empty test suite." I'm not sure how to run them.

**Comment by ligi** [comment]
the tests are now running - lint fails - you can reproduce with:
 ./gradlew lint

you can also look into the Jenkinsfile how my CI executest tests / lint / build

**Comment by ligi** [comment]
:+1: ci is green now - will look through the code soon

File: android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java [hunk 1]
```diff
@@ -141,7 +203,7 @@ public GoMove getParent() {
     }
 
     public GoMove getnextMove(int pos) {
-        return next_move_variations.get(pos);
+        return next_move_variations.size() > pos ? next_move_variations.get(pos) : null;
```
Comment (ligi) [inline android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java]: let' use getOrNull here

File: android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java [hunk 2]
```diff
@@ -19,75 +19,131 @@
...
+        return player;
+    }
+
+    public void setPlayer(byte player) {
+        this.player = player;
+    }
+
     public boolean hasNextMove() {
-        return (next_move_variations.size() > 0);
+        return !next_move_variations.isEmpty();
```
Comment (gthazmatt) [inline android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java]: List doesn't have an isNotEmpty method (though I wish it did). Are you thinking of some utility method like SomeUtil.isNotEmpty(list)?

File: android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java (hunk 1 above)
Comment (gthazmatt) [inline android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java]: I'm not sure what you mean. Can you elaborate?

Comment (ligi) [inline android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java]: Ah I thought we are already in kotlin world here - there getOrNull exists ..

File: android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java (hunk 2 above)
Comment (ligi) [inline android/src/main/java/org/ligi/gobandroid_hd/logic/GoMove.java]: Ah I thought we are already in kotlin world here - there isNotEmpty exists ..

**Comment by ligi** [comment]
Thanks!

**Comment by gthazmatt** [comment]
I think https://github.com/gthazmatt is fine.

**Comment by ligi** [comment]
Released 3.5.5 to beta and you are now in the credits - thanks for your work!

**Pull request merged**